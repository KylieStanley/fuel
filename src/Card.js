import React from 'react'
import './styles/main.scss';

const Card = () => {
	return (
		<div className="card">
		</div>
		
	)
}

export default Card