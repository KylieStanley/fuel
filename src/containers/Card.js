import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { Link, withRouter } from 'react-router-dom'
import { addFavorite, removeFavorite, selectCard } from '../actions'
import '../styles/main.scss'


export class Card extends Component {
	toggleFavorite = () => {
		const {
			favorites, recipe, removeFavorite, addFavorite
		} = this.props

		const match = favorites.find(favorite => favorite.url === recipe.url)

		if (match) {
			removeFavorite(recipe)
		} else {
			addFavorite(recipe)
		}
	}

	render() {
		const { favorites, recipe, selectCard } = this.props
		const favorite = favorites.find(favorite => favorite.url === recipe.url)
		const favoriteClass = favorite ? <i className="favorite fas fa-heart" /> : <i className="far fa-heart" />

		return (
			<div className="card">
				<div className="image" style={{ backgroundImage: `url(${recipe.image})` }} />
				<div className="recipe-info">
					<Link to={`${this.props.match.url}/modal`} onClick={selectCard.bind(null, recipe)}>
						<h3 className="recipe-name">{recipe.name}</h3>
					</Link>
					<div className="card-bottom">
						<button type="button" className="fave-btn" onClick={this.toggleFavorite}>
							{favoriteClass}
						</button>
						{recipe.dietLabel.map(label => <p>{label}</p>)}
					</div>
				</div>
			</div>
		)
	}
}

export const mapStateToProps = (state) => ({
	favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
	addFavorite: (recipe) => dispatch(addFavorite(recipe)),
	removeFavorite: (recipe) => dispatch(removeFavorite(recipe)),
	selectCard: (recipe) => dispatch(selectCard(recipe))
})

Card.propTypes = {
	addFavorite: PropTypes.func.isRequired,
	removeFavorite: PropTypes.func.isRequired,
	selectCard: PropTypes.func.isRequired,
	favorites: PropTypes.arrayOf(PropTypes.object).isRequired,
	recipe: PropTypes.shape({
		name: PropTypes.string,
		yield: PropTypes.number,
		calories: PropTypes.number,
		dietLabel: PropTypes.array,
		url: PropTypes.string,
		image: PropTypes.string,
		source: PropTypes.string,
		ingredients: PropTypes.array
	}).isRequired
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Card))
