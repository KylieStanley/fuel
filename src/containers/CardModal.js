import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { addIngredients } from '../actions'
import '../styles/main.scss'

export const CardModal = (props) => {
	const { recipe } = props
	const ingredientList = recipe.ingredients.map(ingredient => <li key={ingredient}>{ ingredient }</li>)

	return (
		<div className="expanded-card">
			<h2 className="expanded-title">{props.recipe.name}</h2>
			<div className="expanded-info">
				<h3>Serves: {recipe.yield}</h3>
				<h3>Calories per Serving: {recipe.calories}</h3>
				<h3>Source: {recipe.source}</h3>
			</div>
			<div className="ingredients">
				<h2 className="ingredients-title">Ingredients</h2>
				<ul>
					{ ingredientList }
				</ul>
			</div>
			<div className="link">
				<button className="add-btn" type="button" onClick={props.addIngredients.bind(null, props.recipe.ingredients)}>Add Ingredients to shopping List +</button>
				<a
					href={recipe.url}
					target="_blank"
					rel="noopener noreferrer"
				>Go To Full Recipe
					<i className="fas fa-angle-double-right" />
				</a>
			</div>
		</div>
	)
}

CardModal.propTypes = {
	recipe: PropTypes.shape({
		name: PropTypes.string,
		yield: PropTypes.number,
		calories: PropTypes.number,
		dietLabel: PropTypes.array,
		url: PropTypes.string,
		image: PropTypes.string,
		source: PropTypes.string,
		ingredients: PropTypes.array
	}).isRequired
}

export const mapDispatchToProps = (dispatch) => ({
	addIngredients: (ingredients) => dispatch(addIngredients(ingredients))
})

export default connect(null, mapDispatchToProps)(CardModal)
