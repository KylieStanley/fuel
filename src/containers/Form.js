import React, { Component } from 'react'
import '../styles/main.scss';
import { connect } from 'react-redux'
import { cleanSearchString } from '../helper'
import { apiId, apiKey } from '../apiKey'
import { fetchRecipes } from '../thunks/fetchRecipes'
import PropTypes from 'prop-types'


export class Form extends Component {
	constructor() {
		super()
		this.state = {
			search: '',
			filter: ''
		}
	}

	handleChange = (e) => {
		const { name, value } = e.target
		this.setState({ 
			[name]: value 
		}, () => {
			if (name === 'filter') {
				this.handleSubmit(e)
			}
		})	
	}

	handleSubmit = (e) => {
		const { search, filter } = this.state
		const dietType = filter && `&diet=${filter}`
		e.preventDefault()
		if (search) {
			const searchString = cleanSearchString(search)
			const url = `https://api.edamam.com/search?q=${searchString}&app_id=${apiId}&app_key=${apiKey}&from=0&to=30${dietType}`
			this.props.fetchRecipes(url)
		}
	}

	render() {
		return (
		<form onSubmit={this.handleSubmit}>
		<div>
			<input type="text" name="search" placeholder="Search for Recipes" value={this.state.search} onChange={this.handleChange} />
			<button className="search-btn"><i className="fas fa-search"></i></button>
			</div>
			<div className="select-container">
			<select name="filter" onChange={this.handleChange}>  
				<option value="">Filter by Diet</option>
				<option value="balanced">Balanced</option>
				<option value="low-carb">Low Carb</option>
				<option value="high-protein">High Protein</option>
			</select>
			</div>
		</form>
		)
	}
}

export const mapDispatchToProps = (dispatch) => ({
	fetchRecipes: (url) => dispatch(fetchRecipes(url))
})

Form.propTypes = {
	fetchRecipes: PropTypes.func.isRequired
}

export default connect(null, mapDispatchToProps)(Form)