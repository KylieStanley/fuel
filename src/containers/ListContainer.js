import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

export const ListContainer = (props) => {
	const ingredientsList = props.ingredients.map(ingredient => <li key={ingredient} className="item">{ingredient}</li>)

	return (
		<div className="card-container shopping">
			<ul className="shopping-list">
				{ !ingredientsList.length ? <h4>You have no saved ingredients</h4> : ingredientsList }
			</ul>
		</div>
	)
}

export const mapStateToProps = (state) => ({
	ingredients: state.ingredients
})

ListContainer.propTypes = {
	ingredients: PropTypes.arrayOf(PropTypes.string)
}

export default connect(mapStateToProps)(ListContainer)
