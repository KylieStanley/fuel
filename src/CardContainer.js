import React from 'react'
import Card from './Card'
import './styles/main.scss';


const CardContainer = () => {
	return (
		<div className="card-container">
			<Card />
		</div>
	)
}

export default CardContainer