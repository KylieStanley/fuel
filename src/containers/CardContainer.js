import React from 'react'
import '../styles/main.scss'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import Card from './Card'


export const CardContainer = (props) => {
	const { recipes, favorites, isLoading } = props
	let selectedRecipes = recipes

	if (props.itemType === 'favorites') {
		selectedRecipes = favorites
	}

	const error = !selectedRecipes.length && selectedRecipes === favorites ? <h2>You have no favorites!</h2> : null

	return (
		<div className="card-container">
			{ isLoading ? <img src="../loading.gif" alt="spinning circle" /> : selectedRecipes.map(recipe => <Card key={recipe.url} recipe={recipe} />) }
			{ error }
		</div>
	)
}

export const mapStateToProps = (state) => ({
	isLoading: state.isLoading,
	recipes: state.recipes,
	favorites: state.favorites
})

CardContainer.propTypes = {
	isLoading: PropTypes.bool.isRequired,
	recipes: PropTypes.arrayOf(PropTypes.object).isRequired,
	favorites: PropTypes.arrayOf(PropTypes.object).isRequired
}

export default connect(mapStateToProps)(CardContainer)
