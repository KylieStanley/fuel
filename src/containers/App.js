import React, { Component } from 'react';
import '../styles/main.scss';
import { Route, withRouter } from 'react-router-dom'
import Splash from '../components/Splash'
import List from '../components/List'
import { CardModal } from '../components/CardModal'
import Main from './Main'
import { ModalContainer, ModalRoute } from 'react-router-modal';
import { connect } from 'react-redux'
import 'react-router-modal/css/react-router-modal.css'


export const App = (props) => {
  return (
    <div className="splash">
      <div className="overlay" />
      <div className="splash-green" />
      <Route exact path='/' component={ Splash } />
      <Route path='/main' component={ Main } />
      <Route path="/favorites" render={() => <List itemType="favorites" />} />
      <Route path="/shopping" render={() => <List itemType="shopping" />} />
      <ModalRoute component={ CardModal } path='*/modal' parentPath={`${props.history.goBack}`} props={{ recipe: props.selectedCard }} />
      <ModalContainer />
    </div>
  );
}

export const mapStateToProps = (state) => ({
  selectedCard: state.selectedCard
})

export default withRouter(connect(mapStateToProps)(App));

