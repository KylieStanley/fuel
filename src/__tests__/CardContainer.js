import React from 'react'
import { CardContainer, mapStateToProps } from '../containers/CardContainer'
import ReactDOM from 'react-dom';
import Card  from '../containers/Card'
import { connect } from 'react-redux'
import { shallow } from 'enzyme'

describe('CardContainer', () => {
	let wrapper
	let mockRecipes = [{ name: 'chicken' }, { name: 'salmon' }]
	let mockFavorites = [{ name: 'chicken' }]

	beforeEach(() => {
		wrapper = shallow(<CardContainer 
			recipes={ mockRecipes } 
			favorites={ mockFavorites } 
			itemType={ 'recipes' }
		/>)
	})

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('should render recipes if on the main page', () => {
		expect(wrapper.find(Card).length).toEqual(2)
	})

	it('should render favorites if on the favorites page', () => {
		wrapper = shallow(<CardContainer 
			recipes={ mockRecipes } 
			favorites={ mockFavorites } 
			itemType={ 'favorites' }
		/>)
		expect(wrapper.find(Card).length).toEqual(1)
	})

	describe('mapStateToProps', () => {
		it('should return an array of recipes and array of favorites', () => {
			const mockState = {
				recipes: { name: 'chicken' },
				favorites: { name: 'salmon'},
				otherstate: 'item in state'
			}
			const expected = { recipes: mockState.recipes, favorites: mockState.favorites }
			const result = mapStateToProps(mockState)
			expect(result).toEqual(expected)
		})
	})
})