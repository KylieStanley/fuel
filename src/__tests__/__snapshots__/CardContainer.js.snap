// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardContainer
    favorites={
      Array [
        Object {
          "name": "chicken",
        },
      ]
    }
    isLoading={false}
    itemType="recipes"
    recipes={
      Array [
        Object {
          "name": "chicken",
        },
        Object {
          "name": "salmon",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <withRouter(Connect(Card))
            recipe={
              Object {
                "name": "chicken",
              }
            }
          />,
          <withRouter(Connect(Card))
            recipe={
              Object {
                "name": "salmon",
              }
            }
          />,
        ],
        null,
      ],
      "className": "card-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "recipe": Object {
            "name": "chicken",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "recipe": Object {
            "name": "salmon",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <withRouter(Connect(Card))
              recipe={
                Object {
                  "name": "chicken",
                }
              }
            />,
            <withRouter(Connect(Card))
              recipe={
                Object {
                  "name": "salmon",
                }
              }
            />,
          ],
          null,
        ],
        "className": "card-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "recipe": Object {
              "name": "chicken",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "recipe": Object {
              "name": "salmon",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
